File: platform_sims/scripts/gaussianSensorPlatform.py
#!/usr/bin/env python

import rospy #include <ros/ros.h> cpp equivalent
from mavros_msgs.msg import State #include <mavros_msgs/State.h>
from geometry_msgs.msg import PoseStamped
from quadnodes.msg import gaussian

from math import cos, sin, pi, acos, sqrt, exp
import numpy as np

##################
# Global variables
##################

global Robot1_pose
Robot1_pose = PoseStamped()

tmpMsg = gaussian()
# global pose_cb_flag
# pose_cb_flag = True

##################
# Functions
##################

def gaussFunc(xFunc, yFunc, zFunc, QFunc, vFunc, DyFunc, DzFunc):
    con = (QFunc/(4 * pi * xFunc * sqrt(DyFunc*DzFunc))) * exp( -vFunc/(4*xFunc) * ((yFunc**2)/DyFunc + (zFunc**2)/DzFunc))
    return con * 1000 # convert from kg/m^3 to ppm

def getReading(xRobotDef, yRobotDef, thetaFunc, xPlumeFunc, yPlumeFunc, zFunc, QFunc, vFunc, DyFunc, DzFunc):
    # Rotate frame

    Stheta = sin(thetaFunc)
    Ctheta = cos(thetaFunc)

    xRobotRotated = (Ctheta  * xRobotDef + Stheta * yRobotDef + -Ctheta * xPlumeFunc - Stheta * yPlumeFunc)
    yRobotRotated = (-Stheta * xRobotDef + Ctheta * yRobotDef +  Stheta * xPlumeFunc - Ctheta * yPlumeFunc)

    if xRobotRotated <= 0:
        reading = 0
    else:
        reading = gaussFunc(xRobotRotated,yRobotRotated,zFunc,QFunc,vFunc,DyFunc,DzFunc)

    return reading

##################
# Callbacks
##################

def Robot1_pose_cb(Robot1_pose_msg):
    global Robot1_pose
    Robot1_pose = Robot1_pose_msg

##################
# Main
##################

def main():
    rospy.init_node('gaussianPlumeSensor')
    rate = rospy.Rate(50)

    # Plume parameters
    QPlume        = rospy.get_param("/QPlume")                   #  kg/s     release rate
    vPlume        = rospy.get_param("/vPlume")              #  m/s      velocity
    DyPlume       = rospy.get_param("/DyPlume")             #  m        diffusion along y
    DzPlume       = rospy.get_param("/DzPlume")             #  m        diffusion along z

    # Plume orientation translation then rotation
    xPlume     = rospy.get_param("/xPlume")       #  m
    yPlume     = rospy.get_param("/yPlume")       #  m
    zPlume     = rospy.get_param("/zPlume")       #  m
    thetaPlume = rospy.get_param("/thetaPlume")   #  degrees
    careAboutZ = rospy.get_param("/careAboutZ")   #  bool
    sensorRate = rospy.get_param("/sensorRate")   #  Hz

    RobotID    = rospy.get_param("gaussianSensorPlatform/RobotID")   #  Hz

    fullStringName = "/mocap_node/Robot_" + str(int(RobotID)) + "/pose"

    rospy.Subscriber(fullStringName, PoseStamped, Robot1_pose_cb)
    pub = rospy.Publisher("gaussianReading", gaussian, queue_size=10)

    thetaPlume = thetaPlume * (pi/180) # convert to radians

    rate = rospy.Rate(sensorRate)

    while not rospy.is_shutdown():
        # Get reading

        # ppm
        if careAboutZ:
            tmpMsg.ppm = getReading(Robot1_pose.pose.position.x, Robot1_pose.pose.position.y, thetaPlume, xPlume, yPlume, Robot1_pose.pose.position.z - zPlume, QPlume, vPlume, DyPlume, DzPlume)

        else:
            tmpMsg.ppm = getReading(Robot1_pose.pose.position.x, Robot1_pose.pose.position.y, thetaPlume, xPlume, yPlume, 0, QPlume, vPlume, DyPlume, DzPlume)
           #  tmpMsg.ppm = getReading(Robot1_pose.pose.position.x, Robot1_pose.pose.position.y, thetaPlume, xPlume, yPlume, 0, QPlume, vPlume, DyPlume, DzPlume) + np.random.normal(0, 1)

        tmpMsg.x = Robot1_pose.pose.position.x
        tmpMsg.y = Robot1_pose.pose.position.y
        tmpMsg.z = Robot1_pose.pose.position.z

        # Publish message
        pub.publish(tmpMsg);

        rate.sleep()


if __name__ == '__main__':
    try:
        main()
    except rospy.ROSInterruptException:
        pass

File: platform_sims/src/gaussianSensorPlatform.cpp
#include <ros/ros.h>
#include <ros/console.h>
#include <geometry_msgs/PoseStamped.h>
#include <quadnodes/gaussian.h>

#include <math.h>
#include <iostream>

// ##################
// # Functions
// ##################

float gaussFunc(float xFunc, float yFunc, float zFunc, float QFunc, float vFunc, float DyFunc, float DzFunc) {
	// returns concentration at x,y,z cordinates in plume frame
	return (QFunc/(4 * M_PI * xFunc * sqrt(DyFunc*DzFunc))) * exp( -vFunc/(4*xFunc) * (pow(yFunc,2)/DyFunc + pow(zFunc,2)/DzFunc))* 1000;
}

float getReading(float xRobotDef, float yRobotDef, float thetaFunc, float xPlumeFunc, float yPlumeFunc, float zFunc, float QFunc, float vFunc, float DyFunc, float DzFunc) {
	float Stheta = sin(thetaFunc);
	float Ctheta = cos(thetaFunc);

	float XplumeFrame = (Ctheta  * xRobotDef + Stheta * yRobotDef + -Ctheta * xPlumeFunc - Stheta * yPlumeFunc);
	float YplumeFrame = (-Stheta * xRobotDef + Ctheta * yRobotDef +  Stheta * xPlumeFunc - Ctheta * yPlumeFunc);

	float reading;

	if (XplumeFrame <= 0){
		reading = 0;
	}
	else{
		reading = gaussFunc(XplumeFrame, YplumeFrame, zFunc, QFunc, vFunc, DyFunc, DzFunc);
	}

	return reading;
}

// ##################
// # Global variables
// ##################

geometry_msgs::PoseStamped robot_pose;
bool pose_cb_flag = true;

float QPlume;
float vPlume;
float DyPlume;
float DzPlume;

float xPlume;
float yPlume;
float zPlume;
float thetaPlume;

float sensorRate;

bool careAboutZ;

// ##################
// # Callbacks
// ##################

void pose_cb(const geometry_msgs::PoseStamped::ConstPtr& robot_pose_tmp){
  robot_pose = *robot_pose_tmp;
  pose_cb_flag = false;
}

// ##################
// # Main function
// ##################

int main(int argc, char **argv){
  ros::init(argc, argv, "gaussianSensor");
  ros::NodeHandle nh;

  ros::Subscriber controller_sub = nh.subscribe<geometry_msgs::PoseStamped>("mocap_node/Robot_1/pose",1, pose_cb);

	ros::Publisher pub = nh.advertise<quadnodes::gaussian>("gaussianReading",10);

  ros::param::get("/QPlume", QPlume);
  ros::param::get("/vPlume", vPlume);
  ros::param::get("/DyPlume", DyPlume);
  ros::param::get("/DzPlume", DzPlume);

  ros::param::get("/xPlume", xPlume);
  ros::param::get("/yPlume", yPlume);
	ros::param::get("/zPlume", zPlume);
  ros::param::get("/thetaPlume", thetaPlume);
	ros::param::get("/careAboutZ", careAboutZ);

	ros::param::get("/sensorRate", sensorRate);

  ros::Rate rate(sensorRate);

	quadnodes::gaussian gaussianReadingMsg;

  while(ros::ok()){
    ros::spinOnce();
    while (pose_cb_flag) {
      ros::spinOnce();
      if (!pose_cb_flag){
        break;
      }
    }

  // Get reading and publish
	if (!careAboutZ){  // Dont care about z height (2D)
  	// ROS_INFO("Robot 1 reading : %f \n", getReading(robot_pose.pose.position.x,robot_pose.pose.position.y, thetaPlume, xPlume, yPlume, 0.0, QPlume, vPlume, DyPlume, DzPlume));
		gaussianReadingMsg.ppm = getReading(robot_pose.pose.position.x,robot_pose.pose.position.y, thetaPlume, xPlume, yPlume, 0.0, QPlume, vPlume, DyPlume, DzPlume);
	}
	else{
		// ROS_INFO("Robot 1 reading : %f \n", getReading(robot_pose.pose.position.x,robot_pose.pose.position.y, thetaPlume, xPlume, yPlume, robot_pose.pose.position.z - zPlume, QPlume, vPlume, DyPlume, DzPlume));
		gaussianReadingMsg.ppm = getReading(robot_pose.pose.position.x,robot_pose.pose.position.y, thetaPlume, xPlume, yPlume, robot_pose.pose.position.z - zPlume, QPlume, vPlume, DyPlume, DzPlume);
	}

	pub.publish(gaussianReadingMsg);

  rate.sleep();
  }

  return 0;

}
